src/Core/Aetherling/Passes/Timing.hs

```diff
@@ -0,0 +1,265 @@
+module Aetherling.Passes.Timing (
+  distributeUtil,
+  RetimeComposeParPolicy,
+  rcpDefault,
+  rcpRetimeReadyValid,
+  retimeComposePar,
+  retimeComposePar'
+) where
+import Aetherling.Operations.AST
+import Aetherling.Operations.Types
+import Aetherling.Operations.Ops
+import Aetherling.Operations.Compose
+import Aetherling.Analysis.Latency
+import Aetherling.Analysis.PortsAndThroughput
+import Aetherling.Passes.MapOps
+import Data.Ratio
+import Data.List
+
+-- | Given an AST, distribute all LogicalUtil ops to leaf ops, so that
+-- each leaf op in the output AST is wrapped (if needed) with a
+-- LogicalUtil whose ratio is the product of the LogicalUtil ratios on
+-- the path from the leaf to the root (except at ReadyValid
+-- boundaries, the ratio is reset to 1). This allows us to see the
+-- real speed of an op without inspecting the whole AST. Named by
+-- analogy to mathematical distribution A(x+y) -> Ax+Ay. (A is the
+-- metaphorical LogicalUtil, x and y the ops).
+--
+-- Note: This will not remove Delay ops, but their delay amount will
+-- be reset to 1. In this function it's infeasible to determine the
+-- actual delay needed (which way to round for fractional ops?). I set
+-- it to 1 instead of leaving delay untouched to avoid a false sense
+-- of security.
+--
+-- Note: SequenceArrayRepack and Register must know their real speeds,
+-- so it's not wrapped in a LogicalUtil. scaleUtil knows how to modify
+-- their speeds properly.
+distributeUtil :: Op -> Op
+distributeUtil = distributeUtilImpl 1
+
+-- Implementation function for distributeUtil; the Ratio Int arg keeps
+-- track of the product of all LogicalUtil ratios in the path to the
+-- root (or to a ReadyValid meta-op).
+distributeUtilImpl :: (Ratio Int) -> Op -> Op
+distributeUtilImpl ratio0 (LogicalUtil ratio1 op) =
+  distributeUtilImpl (ratio0 * ratio1) op
+distributeUtilImpl ratio (ReadyValid op) =
```

Comment:
please get rid of the special case for readyvalid. This is unintuitive as it makes stuff disappear.